persistence/remote/src/main/java/org/infinispan/persistence/remote/upgrade/MigrationTask.java

```diff
@@ -37,46 +51,97 @@
    public Integer call() throws Exception {
-      ComponentRegistry cr = cache.getAdvancedCache().getComponentRegistry();
-      PersistenceManager loaderManager = cr.getComponent(PersistenceManager.class);
-      Set<RemoteStore> stores = loaderManager.getStores(RemoteStore.class);
-      Marshaller marshaller = new MigrationMarshaller();
-      byte[] knownKeys;
       try {
-         knownKeys = marshaller.objectToByteBuffer(MIGRATION_MANAGER_HOT_ROD_KNOWN_KEYS);
-      } catch (Exception e) {
-         throw new CacheException(e);
+         Iterator<RemoteStore> storeIterator = stores.iterator();
+         if (storeIterator.hasNext()) {
+            RemoteStore store = storeIterator.next();
+            RemoteCache<Object, Object> storeCache = store.getRemoteCache();
+            RemoteStoreConfiguration storeConfig = store.getConfiguration();
+            if (!storeConfig.hotRodWrapping()) {
+               throw log.remoteStoreNoHotRodWrapping(cache.getName());
+            }
+            AtomicInteger counter = new AtomicInteger(0);
+            migrateEntriesWithMetadata(storeCache,counter);
+            gracefulShutdown(executorService);
+            return counter.intValue();
+         }
+         return null;
+      } finally {
+         cache.removeListener(listener);
       }
-      Iterator<RemoteStore> storeIterator = stores.iterator();
-      if (storeIterator.hasNext()) {
-         RemoteStore store = storeIterator.next();
-         final RemoteCache<Object, Object> storeCache = store.getRemoteCache();
-         RemoteStoreConfiguration storeConfig = store.getConfiguration();
-         if (!storeConfig.hotRodWrapping()) {
-            throw log.remoteStoreNoHotRodWrapping(cache.getName());
+   }
+
+   private void migrateEntriesWithMetadata(RemoteCache<Object, Object> sourceCache, AtomicInteger counter) {
+      try (CloseableIterator<Map.Entry<Object, MetadataValue<Object>>> iterator = sourceCache.retrieveEntriesWithMetadata(segments, readBatch)) {
+         while (iterator.hasNext()) {
+            Map.Entry<Object, MetadataValue<Object>> entry = iterator.next();
+            if (!Arrays.equals((byte[]) entry.getKey(), ignoredKey)) {
+               MetadataValue<Object> metadataValue = entry.getValue();
+               int lifespan = metadataValue.getLifespan();
+               int maxIdle = metadataValue.getMaxIdle();
+               long version = metadataValue.getVersion();
+               long created = metadataValue.getCreated();
+               long lastUsed = metadataValue.getLastUsed();
+               Metadata metadata = new EmbeddedMetadata.Builder()
+                     .version(new NumericVersion(version))
+                     .lifespan(lifespan, TimeUnit.SECONDS)
+                     .maxIdle(maxIdle, TimeUnit.SECONDS)
+                     .build();
+               InternalMetadataImpl internalMetadata = new InternalMetadataImpl(metadata, created, lastUsed);
+               executorService.submit(() -> {
```

Comment:
No difference noted, plus this would add another knob that would interfere with the existing ones at https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/upgrade/TargetMigrator.java#L25